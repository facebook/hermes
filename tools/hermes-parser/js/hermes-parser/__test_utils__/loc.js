/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow strict-local
 * @format
 */

'use strict';

import type {SourceLocation} from '../../hermes-estree/src/types';

/**
 * Utility for quickly creating source locations inline.
 */
export function loc(
  startLine: number,
  startColumn: number,
  endLine: number,
  endColumn: number,
): SourceLocation {
  return {
    start: {
      line: startLine,
      column: startColumn,
    },
    end: {
      line: endLine,
      column: endColumn,
    },
  };
}
