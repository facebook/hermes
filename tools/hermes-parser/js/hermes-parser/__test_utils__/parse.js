/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow strict-local
 * @format
 */

'use strict';

import type {ESNode} from 'hermes-estree';

import {SimpleTraverser} from '../src/traverse/SimpleTraverser';
import {parse as parseOriginal} from '../src/index';
import {print as printAST} from 'hermes-transform';

// $FlowExpectedError[untyped-import]
import {VISITOR_KEYS as babelVisitorKeys} from '@babel/types';
// $FlowExpectedError[untyped-import]
import generate from '@babel/generator';

const prettierConfig = Object.freeze({
  arrowParens: 'avoid',
  singleQuote: true,
  trailingComma: 'all',
  bracketSpacing: false,
  bracketSameLine: true,
  parser: 'hermes',
});

export const parse: typeof parseOriginal = (source, options) => {
  return parseOriginal(source, {flow: 'all', ...options});
};

export function parseForSnapshot(
  source: string,
  {
    babel,
    preserveRange,
    enableExperimentalComponentSyntax,
  }: {
    preserveRange?: boolean,
    babel?: boolean,
    enableExperimentalComponentSyntax?: boolean,
  } = {},
): mixed {
  const parseOpts = {
    enableExperimentalComponentSyntax:
      enableExperimentalComponentSyntax ?? false,
  };
  if (babel === true) {
    return cleanASTForSnapshot(
      parse(source, {
        babel: true,
        ...parseOpts,
      }).program,
      {babel, preserveRange},
    );
  }

  return cleanASTForSnapshot(parse(source, parseOpts), {
    babel,
    preserveRange,
  });
}

export async function printForSnapshot(
  source: string,
  {
    babel,
    enableExperimentalComponentSyntax,
  }: {
    babel?: boolean,
    enableExperimentalComponentSyntax?: boolean,
  } = {},
): Promise<string> {
  const parseOpts = {
    enableExperimentalComponentSyntax:
      enableExperimentalComponentSyntax ?? false,
  };
  if (babel === true) {
    const ast = parse(source, {
      babel: true,
      ...parseOpts,
    }).program;
    return generate(ast).code;
  }

  const ast = parse(source, parseOpts);
  const output = await printAST(ast, source, prettierConfig);
  return output.trim();
}

export function cleanASTForSnapshot(
  ast: ESNode,
  options?: {preserveRange?: boolean, babel?: boolean},
): mixed {
  SimpleTraverser.traverse(ast, {
    enter(node) {
      // $FlowExpectedError[cannot-write]
      delete node.loc;
      // $FlowExpectedError[cannot-write]
      delete node.parent;
      if (options?.preserveRange !== true) {
        // $FlowExpectedError[cannot-write]
        delete node.range;
      }

      if (options?.babel === true) {
        // $FlowExpectedError[prop-missing]
        delete node.start;
        // $FlowExpectedError[prop-missing]
        delete node.end;
      }
    },
    leave() {},
    visitorKeys:
      options?.babel === true
        ? {...babelVisitorKeys, BigIntLiteralTypeAnnotation: []}
        : null,
  });

  if (ast.type === 'Program') {
    return {
      type: 'Program',
      body: ast.body,
    };
  }

  return ast;
}
