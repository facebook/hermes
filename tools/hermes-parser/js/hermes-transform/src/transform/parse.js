/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow strict-local
 * @format
 */

'use strict';

import type {Program} from 'hermes-estree';
import type {ScopeManager} from 'hermes-eslint';

import {attachComments} from './comments/comments';
import {parseForESLint} from 'hermes-eslint';

export type ParseResult = {
  ast: Program,
  scopeManager: ScopeManager,
  code: string,
};

export async function parse(code: string): Promise<ParseResult> {
  const {ast, scopeManager} = parseForESLint(code, {
    sourceType: 'module',
    enableExperimentalComponentSyntax: true,
  });

  // Don't include the docblock comment in the comment list as we don't want to attach it
  // as it should be maintained at the top of the file as nodes are moved around.
  let comments = ast.comments;
  if (ast.docblock != null && ast.docblock.comment === comments[0]) {
    const [first, ...nonDocblockComments] = comments;

    // Since we will not be attaching this comment automatically we need to add the
    // properties prettier expects for printing.
    // $FlowExpectedError[prop-missing]
    first.placement = 'endOfLine';
    // $FlowExpectedError[prop-missing]
    first.leading = true;
    // $FlowExpectedError[prop-missing]
    first.trailing = false;
    // $FlowExpectedError[prop-missing]
    first.printed = false;

    comments = nonDocblockComments;
  }

  // attach comments before mutation. this will ensure that as nodes are
  // cloned / moved around - comments remain in the correct place with respect to the node
  attachComments(comments, ast, code);

  return {ast, scopeManager, code};
}
