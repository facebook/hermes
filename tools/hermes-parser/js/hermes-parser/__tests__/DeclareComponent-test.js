/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow strict-local
 * @format
 */

import type {AlignmentCase} from '../__test_utils__/alignment-utils';

import {
  expectBabelAlignment,
  expectEspreeAlignment,
} from '../__test_utils__/alignment-utils';
import {parseForSnapshot} from '../__test_utils__/parse';

const parserOpts = {enableExperimentalComponentSyntax: true};

describe('DeclareComponent', () => {
  const testCase: AlignmentCase = {
    code: `
      declare component Foo();
    `,
    espree: {
      expectToFail: 'espree-exception',
      expectedExceptionMessage: 'Unexpected token component',
    },
    babel: {
      expectToFail: 'babel-exception',
      expectedExceptionMessage: 'Unexpected token',
    },
  };

  test('ESTree', () => {
    expect(parseForSnapshot(testCase.code, parserOpts)).toMatchInlineSnapshot(`
      {
        "body": [
          {
            "id": {
              "name": "Foo",
              "optional": false,
              "type": "Identifier",
              "typeAnnotation": null,
            },
            "params": [],
            "rendersType": null,
            "rest": null,
            "type": "DeclareComponent",
            "typeParameters": null,
          },
        ],
        "type": "Program",
      }
    `);
    expectEspreeAlignment(testCase);
  });

  test('Babel', () => {
    expect(parseForSnapshot(testCase.code, {babel: true, ...parserOpts}))
      .toMatchInlineSnapshot(`
      {
        "body": [
          {
            "id": {
              "name": "Foo",
              "optional": false,
              "type": "Identifier",
              "typeAnnotation": {
                "type": "AnyTypeAnnotation",
              },
            },
            "type": "DeclareVariable",
          },
        ],
        "type": "Program",
      }
    `);
    expectBabelAlignment(testCase);
  });
});
