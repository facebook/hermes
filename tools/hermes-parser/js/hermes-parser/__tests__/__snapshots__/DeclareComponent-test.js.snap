// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeclareComponent Basic Babel 1`] = `
{
  "body": [
    {
      "id": {
        "name": "Foo",
        "type": "Identifier",
        "typeAnnotation": {
          "type": "TypeAnnotation",
          "typeAnnotation": {
            "type": "AnyTypeAnnotation",
          },
        },
      },
      "type": "DeclareVariable",
    },
  ],
  "type": "Program",
}
`;

exports[`DeclareComponent Basic ESTree 1`] = `
{
  "body": [
    {
      "id": {
        "name": "Foo",
        "optional": false,
        "type": "Identifier",
        "typeAnnotation": null,
      },
      "params": [],
      "rendersType": null,
      "rest": null,
      "type": "DeclareComponent",
      "typeParameters": null,
    },
  ],
  "type": "Program",
}
`;

exports[`DeclareComponent Export Babel 1`] = `
{
  "body": [
    {
      "declaration": {
        "id": {
          "name": "Foo",
          "type": "Identifier",
          "typeAnnotation": {
            "type": "TypeAnnotation",
            "typeAnnotation": {
              "type": "AnyTypeAnnotation",
            },
          },
        },
        "type": "DeclareVariable",
      },
      "default": false,
      "source": null,
      "specifiers": [],
      "type": "DeclareExportDeclaration",
    },
    {
      "declaration": {
        "id": {
          "name": "Bar",
          "type": "Identifier",
          "typeAnnotation": {
            "type": "TypeAnnotation",
            "typeAnnotation": {
              "type": "AnyTypeAnnotation",
            },
          },
        },
        "type": "DeclareVariable",
      },
      "default": true,
      "source": null,
      "specifiers": [],
      "type": "DeclareExportDeclaration",
    },
  ],
  "type": "Program",
}
`;

exports[`DeclareComponent Export ESTree 1`] = `
{
  "body": [
    {
      "declaration": {
        "id": {
          "name": "Foo",
          "optional": false,
          "type": "Identifier",
          "typeAnnotation": null,
        },
        "params": [],
        "rendersType": null,
        "rest": null,
        "type": "DeclareComponent",
        "typeParameters": null,
      },
      "default": false,
      "source": null,
      "specifiers": [],
      "type": "DeclareExportDeclaration",
    },
    {
      "declaration": {
        "id": {
          "name": "Bar",
          "optional": false,
          "type": "Identifier",
          "typeAnnotation": null,
        },
        "params": [],
        "rendersType": null,
        "rest": null,
        "type": "DeclareComponent",
        "typeParameters": null,
      },
      "default": true,
      "source": null,
      "specifiers": [],
      "type": "DeclareExportDeclaration",
    },
  ],
  "type": "Program",
}
`;

exports[`DeclareComponent Params and renders Babel 1`] = `
{
  "body": [
    {
      "id": {
        "name": "Foo",
        "type": "Identifier",
        "typeAnnotation": {
          "type": "TypeAnnotation",
          "typeAnnotation": {
            "type": "AnyTypeAnnotation",
          },
        },
      },
      "type": "DeclareVariable",
    },
  ],
  "type": "Program",
}
`;

exports[`DeclareComponent Params and renders ESTree 1`] = `
{
  "body": [
    {
      "id": {
        "name": "Foo",
        "optional": false,
        "type": "Identifier",
        "typeAnnotation": null,
      },
      "params": [
        {
          "name": {
            "name": "bar",
            "optional": false,
            "type": "Identifier",
            "typeAnnotation": null,
          },
          "optional": false,
          "type": "ComponentTypeParameter",
          "typeAnnotation": {
            "type": "StringTypeAnnotation",
          },
        },
      ],
      "rendersType": {
        "operator": "renders",
        "type": "TypeOperator",
        "typeAnnotation": {
          "id": {
            "name": "Foo",
            "optional": false,
            "type": "Identifier",
            "typeAnnotation": null,
          },
          "type": "GenericTypeAnnotation",
          "typeParameters": null,
        },
      },
      "rest": null,
      "type": "DeclareComponent",
      "typeParameters": null,
    },
  ],
  "type": "Program",
}
`;

exports[`DeclareComponent Rest param Babel 1`] = `
{
  "body": [
    {
      "id": {
        "name": "Foo",
        "type": "Identifier",
        "typeAnnotation": {
          "type": "TypeAnnotation",
          "typeAnnotation": {
            "type": "AnyTypeAnnotation",
          },
        },
      },
      "type": "DeclareVariable",
    },
  ],
  "type": "Program",
}
`;

exports[`DeclareComponent Rest param ESTree 1`] = `
{
  "body": [
    {
      "id": {
        "name": "Foo",
        "optional": false,
        "type": "Identifier",
        "typeAnnotation": null,
      },
      "params": [],
      "rendersType": null,
      "rest": {
        "name": {
          "name": "rest",
          "optional": false,
          "type": "Identifier",
          "typeAnnotation": null,
        },
        "optional": false,
        "type": "ComponentTypeParameter",
        "typeAnnotation": {
          "callProperties": [],
          "exact": false,
          "indexers": [],
          "inexact": false,
          "internalSlots": [],
          "properties": [
            {
              "key": {
                "name": "bar",
                "optional": false,
                "type": "Identifier",
                "typeAnnotation": null,
              },
              "kind": "init",
              "method": false,
              "optional": false,
              "proto": false,
              "static": false,
              "type": "ObjectTypeProperty",
              "value": {
                "type": "StringTypeAnnotation",
              },
              "variance": null,
            },
          ],
          "type": "ObjectTypeAnnotation",
        },
      },
      "type": "DeclareComponent",
      "typeParameters": null,
    },
  ],
  "type": "Program",
}
`;
